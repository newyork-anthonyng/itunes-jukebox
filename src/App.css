input {
  display: block;
  margin-bottom: 1rem;
  margin-top: 1rem;
}

ul {
  list-style: none;
  padding: 0;
}

li {
  cursor: pointer;
  position: relative;
}

li:hover:after {
  position: absolute;
  content: "ðŸ‘ˆ";
  font-size: 50px;
  top: 50%;
  transform: translateY(-50%);
}

li img {
  margin-right: 16px;
}