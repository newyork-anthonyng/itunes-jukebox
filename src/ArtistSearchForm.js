import React from "react";
import PropTypes from "prop-types";

const ArtistSearchForm = ({ onSubmit, onChange, value }) => (
  <form onSubmit={onSubmit}>
    <label>
      Search albums for artist
      <input
        type="text"
        placeholder="Enter artist name"
        value={value}
        onChange={onChange}
      />
    </label>

    <button type="submit">Search</button>
  </form>
);

ArtistSearchForm.propTypes = {
  onSubmit: PropTypes.func,
  onChange: PropTypes.func,
  value: PropTypes.string
};

export default ArtistSearchForm;
