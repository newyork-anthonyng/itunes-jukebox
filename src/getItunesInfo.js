import axios from "axios";

const constructAlbumUrl = artistName => {
    // https://developer.apple.com/library/content/documentation/AudioVideo/Conceptual/iTuneSearchAPI/Searching.html#//apple_ref/doc/uid/TP40017632-CH5-SW2
    return `https://itunes.apple.com/search?term=${artistName.split(" ").join("+")}&entity=album`;
}

const getAlbumsForArtist = artistName => {
    return axios.get(constructAlbumUrl(artistName));
};

const formatAlbumsInfo = axiosResponse => {
    return {
        albums: axiosResponse.data.results.map(formatAlbumInfo)
    };
};

const formatAlbumInfo = ({ collectionId, collectionName, artistName, artworkUrl100, collectionViewUrl }) => ({
    id: collectionId,
    artistName,
    albumArt: artworkUrl100,
    albumName: collectionName,
    itunesLink: collectionViewUrl
});

export {
    getAlbumsForArtist,
    formatAlbumsInfo
};