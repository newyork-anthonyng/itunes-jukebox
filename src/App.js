import React, { Component } from "react";
import "./App.css";
import ArtistSearchForm from "./ArtistSearchForm";
import AlbumList from "./AlbumList";
import { getAlbumsForArtist, formatAlbumsInfo } from "./getItunesInfo";

// TODO: Add styling

class App extends Component {
  constructor(props) {
    super(props);

    this.state = {
      textInput: "",
      albums: [],
      error: null,
      isLoading: false
    };
  }

  handleInputChange = e => {
    this.setState({
      textInput: e.target.value,
      error: null
    });
  };

  handleSubmit = e => {
    e.preventDefault();

    if (this.state.isLoading) {
      return;
    }

    if (this.state.textInput.trim() === "") {
      return this.setState({
        error: "Please enter an artist name",
        isLoading: false
      });
    }

    this.setState({
      error: null,
      isLoading: true
    });

    return getAlbumsForArtist(this.state.textInput)
      .then(response => {
        const result = formatAlbumsInfo(response);

        this.setState({
          albums: (result && result.albums) || [],
          isLoading: false
        });
      })
      .catch(e => {
        console.error(e);

        this.setState({
          error:
            "Whoops. An error occurred when trying to get album information.",
          isLoading: false
        });
      });
  };

  renderError = () => {
    return <h2 role="alert">{this.state.error}</h2>;
  };

  render() {
    const { textInput, albums, isLoading, error } = this.state;

    return (
      <div>
        {error && this.renderError()}
        <ArtistSearchForm
          onSubmit={this.handleSubmit}
          onChange={this.handleInputChange}
          value={textInput}
        />

        {isLoading && <span>🐾</span>}

        {albums.length > 0 && (
          <AlbumList artistName={albums[0].artistName} albums={albums} />
        )}
      </div>
    );
  }
}

export default App;
