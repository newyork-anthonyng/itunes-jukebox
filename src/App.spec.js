import React from "react";
import { shallow } from "enzyme";
import toJSON from "enzyme-to-json";
import App from "./App";
import ArtistSearchForm from "./ArtistSearchForm";
import { getAlbumsForArtist, formatAlbumsInfo } from "./getItunesInfo";
jest.mock("./getItunesInfo", () => ({
  getAlbumsForArtist: jest.fn(() =>
    Promise.resolve({
      albums: []
    })
  ),
  formatAlbumsInfo: jest.fn()
}));

beforeEach(() => {
  getAlbumsForArtist.mockClear();
});

it("should render correctly", () => {
  const wrapper = shallow(<App />);

  expect(toJSON(wrapper)).toMatchSnapshot();
});

it("should render correctly with albums", () => {
  const wrapper = shallow(<App />);
  wrapper.setState({
    albums: [
      {
        albumArt:
          "https://is1-ssl.mzstatic.com/image/thumb/Music3/v4/fb/af/59/fbaf5908-0839-abc6-9f6a-bc7cc5b84f27/source/100x100bb.jpg",
        albumName: "American Beauty / American Psycho",
        artistName: "Fall Out Boy",
        id: 948754194,
        itunesLink:
          "https://itunes.apple.com/us/album/american-beauty-american-psycho/948754194?uo=4"
      }
    ]
  });

  expect(toJSON(wrapper)).toMatchSnapshot();
});

it("should render correctly when there is an error", () => {
  const wrapper = shallow(<App />);
  wrapper.setState({ error: "Uh oh. Something went wrong" });

  expect(toJSON(wrapper)).toMatchSnapshot();
});

it("should should update correctly when updating input", () => {
  const wrapper = shallow(<App />);
  wrapper.setState({ error: "Uh oh" });
  const searchForm = wrapper.find(ArtistSearchForm);

  searchForm.props().onChange({ target: { value: "Michael" } });

  wrapper.update();

  expect(wrapper.find(ArtistSearchForm).props().value).toEqual("Michael");
  expect(toJSON(wrapper)).toMatchSnapshot();
});

describe("Submitting form", () => {
  it("should call API response when submitting form", () => {
    const wrapper = shallow(<App />);
    wrapper.setState({ textInput: "Michael Jackson" });

    const searchForm = wrapper.find(ArtistSearchForm);
    const mockEvent = { preventDefault: jest.fn() };
    searchForm.props().onSubmit(mockEvent);

    expect(mockEvent.preventDefault).toHaveBeenCalledTimes(1);
    expect(getAlbumsForArtist).toHaveBeenCalledTimes(1);
    expect(getAlbumsForArtist.mock.calls[0]).toMatchSnapshot();
  });

  it("should update correctly when receiving correct API response", () => {
    let apiPromise;
    formatAlbumsInfo.mockImplementation(() => ({
      albums: [
        {
          albumArt:
            "https://is1-ssl.mzstatic.com/image/thumb/Music3/v4/fb/af/59/fbaf5908-0839-abc6-9f6a-bc7cc5b84f27/source/100x100bb.jpg",
          albumName: "American Beauty / American Psycho",
          artistName: "Fall Out Boy",
          id: 948754194,
          itunesLink:
            "https://itunes.apple.com/us/album/american-beauty-american-psycho/948754194?uo=4"
        }
      ]
    }));
    getAlbumsForArtist.mockImplementation(() => {
      apiPromise = Promise.resolve({});
      return apiPromise;
    });
    const wrapper = shallow(<App />);
    wrapper.setState({ textInput: "Michael Jackson" });

    const searchForm = wrapper.find(ArtistSearchForm);
    const mockEvent = { preventDefault: jest.fn() };
    searchForm.props().onSubmit(mockEvent);

    return apiPromise.then(() => {
      wrapper.update();

      expect(toJSON(wrapper)).toMatchSnapshot();
      expect(wrapper.state()).toMatchSnapshot();
    });
  });

  it("should update correctly when there is an error from API response", () => {
    let apiPromise;
    getAlbumsForArtist.mockImplementation(() => {
      apiPromise = Promise.reject("Something bad happened");
      return apiPromise;
    });
    const wrapper = shallow(<App />);
    wrapper.setState({ textInput: "Michael Jackson" });

    const searchForm = wrapper.find(ArtistSearchForm);
    const mockEvent = { preventDefault: jest.fn() };
    searchForm.props().onSubmit(mockEvent);

    return apiPromise.then().catch(() => {
      wrapper.update();

      expect(toJSON(wrapper)).toMatchSnapshot();
      expect(wrapper.state()).toMatchSnapshot();
    });
  });

  it("should update correctly when submitting form and App is loading", () => {
    const wrapper = shallow(<App />);
    wrapper.setState({ textInput: "Michael Jackson", isLoading: true });

    const searchForm = wrapper.find(ArtistSearchForm);
    const mockEvent = { preventDefault: jest.fn() };
    searchForm.props().onSubmit(mockEvent);

    expect(getAlbumsForArtist).not.toHaveBeenCalled();
  });

  it("should update correctly when submitting form and App text is empty", () => {
    const wrapper = shallow(<App />);
    wrapper.setState({ textInput: " " });

    const searchForm = wrapper.find(ArtistSearchForm);
    const mockEvent = { preventDefault: jest.fn() };
    searchForm.props().onSubmit(mockEvent);

    expect(getAlbumsForArtist).not.toHaveBeenCalled();
  });
});
