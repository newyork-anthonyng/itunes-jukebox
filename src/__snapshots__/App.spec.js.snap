// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Submitting form should call API response when submitting form 1`] = `
Array [
  "Michael Jackson",
]
`;

exports[`Submitting form should update correctly when receiving correct API response 1`] = `
<div>
  <ArtistSearchForm
    onChange={[Function]}
    onSubmit={[Function]}
    value="Michael Jackson"
  />
  <AlbumList
    albums={
      Array [
        Object {
          "albumArt": "https://is1-ssl.mzstatic.com/image/thumb/Music3/v4/fb/af/59/fbaf5908-0839-abc6-9f6a-bc7cc5b84f27/source/100x100bb.jpg",
          "albumName": "American Beauty / American Psycho",
          "artistName": "Fall Out Boy",
          "id": 948754194,
          "itunesLink": "https://itunes.apple.com/us/album/american-beauty-american-psycho/948754194?uo=4",
        },
      ]
    }
    artistName="Fall Out Boy"
  />
</div>
`;

exports[`Submitting form should update correctly when receiving correct API response 2`] = `
Object {
  "albums": Array [
    Object {
      "albumArt": "https://is1-ssl.mzstatic.com/image/thumb/Music3/v4/fb/af/59/fbaf5908-0839-abc6-9f6a-bc7cc5b84f27/source/100x100bb.jpg",
      "albumName": "American Beauty / American Psycho",
      "artistName": "Fall Out Boy",
      "id": 948754194,
      "itunesLink": "https://itunes.apple.com/us/album/american-beauty-american-psycho/948754194?uo=4",
    },
  ],
  "error": null,
  "isLoading": false,
  "textInput": "Michael Jackson",
}
`;

exports[`Submitting form should update correctly when there is an error from API response 1`] = `
<div>
  <h2
    role="alert"
  >
    Whoops. An error occurred when trying to get album information.
  </h2>
  <ArtistSearchForm
    onChange={[Function]}
    onSubmit={[Function]}
    value="Michael Jackson"
  />
</div>
`;

exports[`Submitting form should update correctly when there is an error from API response 2`] = `
Object {
  "albums": Array [],
  "error": "Whoops. An error occurred when trying to get album information.",
  "isLoading": false,
  "textInput": "Michael Jackson",
}
`;

exports[`should render correctly 1`] = `
<div>
  <ArtistSearchForm
    onChange={[Function]}
    onSubmit={[Function]}
    value=""
  />
</div>
`;

exports[`should render correctly when there is an error 1`] = `
<div>
  <h2
    role="alert"
  >
    Uh oh. Something went wrong
  </h2>
  <ArtistSearchForm
    onChange={[Function]}
    onSubmit={[Function]}
    value=""
  />
</div>
`;

exports[`should render correctly with albums 1`] = `
<div>
  <ArtistSearchForm
    onChange={[Function]}
    onSubmit={[Function]}
    value=""
  />
  <AlbumList
    albums={
      Array [
        Object {
          "albumArt": "https://is1-ssl.mzstatic.com/image/thumb/Music3/v4/fb/af/59/fbaf5908-0839-abc6-9f6a-bc7cc5b84f27/source/100x100bb.jpg",
          "albumName": "American Beauty / American Psycho",
          "artistName": "Fall Out Boy",
          "id": 948754194,
          "itunesLink": "https://itunes.apple.com/us/album/american-beauty-american-psycho/948754194?uo=4",
        },
      ]
    }
    artistName="Fall Out Boy"
  />
</div>
`;

exports[`should should update correctly when updating input 1`] = `
<div>
  <ArtistSearchForm
    onChange={[Function]}
    onSubmit={[Function]}
    value="Michael"
  />
</div>
`;
